librz/core/cmd_system.c

```diff
@@ -0,0 +1,217 @@
+		close(fd);
+	}
+	if (!path) {
+		return NULL;
+	}
+
+	Sdb *sdb = sdb_new(NULL, path, 0);
+	if (!sdb) {
+		return NULL;
+	}
+	rz_config_serialize(core->config, sdb);
+	sdb_sync(sdb);
+	sdb_free(sdb);
+
+	return path;
+}
+
+static const char *system_apply_env_var(const char *env, const char *value, const char *arg, RzList *alloc) {
+	char *string = NULL;
+	RzListIter *it = NULL;
+	size_t len = strlen(arg);
+	if (!strstr(arg, env)) {
+		return arg;
+	} else if (strlen(env) == len) {
+		return value;
+	}
+
+	if ((it = rz_list_find_ptr(alloc, (void *)arg))) {
+		string = it->data;
+	} else {
+		string = strdup(arg);
+	}
+
+	string = rz_str_replace(string, env, value, 1);
+	if (it) {
+		it->data = string;
+	} else {
+		rz_list_append(alloc, string);
+	}
+	return string;
+}
+
+static bool system_exec(RzCore *core, int argc, const char **argv, char **output, int *length) {
+	char file_size[32];
+	char core_offset[32];
+	char block_size[32];
+	const char *file_path = rz_config_get(core->config, "file.path");
+	const char *asm_arch = rz_config_get(core->config, "asm.arch");
+	const char *asm_bits = rz_config_get(core->config, "asm.bits");
+	const char *bin_demangle = rz_config_get(core->config, "bin.demangle");
+	const char *bin_lang = rz_config_get(core->config, "bin.lang");
+	const char *cfg_debug = rz_config_get(core->config, "cfg.debug");
+	const char *io_va = rz_config_get(core->config, "io.va");
+	const char *pdb_server = rz_config_get(core->config, "pdb.server");
+	const char *scr_color = rz_config_get(core->config, "scr.color");
+	const char *endian = rz_str_bool(core->rasm->big_endian);
+	char *cfg_path = config_path(core);
+	snprintf(file_size, sizeof(file_size), "%" PFMT64u, core->file ? rz_io_fd_size(core->io, core->file->fd) : 0);
+	snprintf(core_offset, sizeof(core_offset), "%" PFMT64u, core->offset);
+	snprintf(block_size, sizeof(block_size), "%u", core->blocksize);
+
+	const char *envvars[] = {
+		"RZ_FILE",
+		"RZ_SIZE",
+		"RZ_ARCH",
+		"RZ_BITS",
+		"RZ_OFFSET",
+		"RZ_ENDIAN",
+		"RZ_BIN_DEMANGLE",
+		"RZ_BIN_LANG",
+		"RZ_BIN_PDBSERVER",
+		"RZ_IOVA",
+		"RZ_COLOR",
+		"RZ_BSIZE",
+		"RZ_DEBUG",
+		"RZ_CONFIG"
+	};
+	const char *envvals[] = {
+		file_path,
+		file_size,
+		asm_arch,
+		asm_bits,
+		core_offset,
+		endian,
+		bin_demangle,
+		bin_lang,
+		pdb_server,
+		io_va,
+		scr_color,
+		block_size,
+		cfg_debug,
+		cfg_path ? cfg_path : ""
+	};
+
+	RzList *alloc = rz_list_newf(free);
+	if (!alloc) {
+		RZ_LOG_ERROR("Cannot allocate list of allocated strings.\n");
+		rz_file_rm(cfg_path);
+		free(cfg_path);
+		return false;
+	}
+
+	const char **args = RZ_NEWS0(const char *, argc);
+	if (!args) {
+		RZ_LOG_ERROR("Cannot allocate list of args.\n");
+		rz_list_free(alloc);
+		rz_file_rm(cfg_path);
+		free(cfg_path);
+		return false;
+	}
+
+	if (!rz_subprocess_init()) {
+		RZ_LOG_ERROR("Cannot initialize subprocess.\n");
+		rz_list_free(alloc);
+		rz_file_rm(cfg_path);
+		free(cfg_path);
+		free(args);
+		return false;
```

Comment:
put this at the end of the function and use `goto`, it will simplify errors handling and make sure things are done in the right order on each error.